api/src/main/java/org/pmiops/workbench/cohorts/CohortMaterializationService.java

```diff
@@ -0,0 +1,81 @@
+import org.pmiops.workbench.cohortbuilder.ParticipantCounter;
+import org.pmiops.workbench.db.model.CdrVersion;
+import org.pmiops.workbench.exceptions.BadRequestException;
+import org.pmiops.workbench.model.CohortStatus;
+import org.pmiops.workbench.model.MaterializeCohortResponse;
+import org.pmiops.workbench.model.SearchRequest;
+import org.pmiops.workbench.utils.PaginationToken;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Service;
+
+@Service
+public class CohortMaterializationService {
+
+  private final BigQueryService bigQueryService;
+  private final ParticipantCounter participantCounter;
+
+  @Autowired
+  public CohortMaterializationService(BigQueryService bigQueryService,
+        ParticipantCounter participantCounter) {
+    this.bigQueryService = bigQueryService;
+    this.participantCounter = participantCounter;
+  }
+
+
+  public MaterializeCohortResponse materializeCohort(CdrVersion cdrVersion, String cohortSpec,
+      List<CohortStatus> statusFilter, int pageSize, String paginationToken) {
+    SearchRequest searchRequest;
+    try {
+      searchRequest = new Gson().fromJson(cohortSpec, SearchRequest.class);
+    } catch (JsonSyntaxException e) {
+      throw new BadRequestException("Invalid cohort spec");
+    }
+    long offset = 0L;
+    // TODO: add CDR version ID here
+    Object[] paginationParameters = new Object[] { cohortSpec, statusFilter };
+    if (paginationToken != null) {
+      PaginationToken token = PaginationToken.fromBase64(paginationToken);
+      if (token.matchesParameters(paginationParameters)) {
+        offset = token.getOffset();
+      } else {
+        throw new BadRequestException(
+            String.format("Use of pagination token %s with new parameter values", paginationToken));
+      }
+    }
+    int limit = pageSize + 1;
+    // TODO: use CDR version, statusFilter here
+    QueryResult result = bigQueryService.executeQuery(bigQueryService.filterBigQueryConfig(
+        participantCounter.buildParticipantIdQuery(searchRequest, limit, offset)));
+    MaterializeCohortResponse response = new MaterializeCohortResponse();
+    Map<String, Integer> rm = bigQueryService.getResultMapper(result);
+    int numResults = 0;
+    boolean hasMoreResults = false;
+    for (List<FieldValue> row : result.iterateAll()) {
+      long personId = bigQueryService.getLong(row, rm.get("person_id"));
+      Map<String, Object> resultMap = new HashMap<>(1);
+      resultMap.put("person_id", personId);
+      response.addResultsItem(resultMap);
+      numResults++;
+      if (numResults == pageSize) {
+        hasMoreResults = true;
+        break;
+      }
+    }
+    if (hasMoreResults) {
+      // TODO: consider pagination based on cursor / values rather than offset
+      PaginationToken token = PaginationToken.of(offset + pageSize, paginationParameters);
+      response.setNextPageToken(token.toBase64());
+    }
+    return response;
+  }
+}
```

Comment:
Seems fine. Should we have some unit test for this class?

Reply:
Yep, adding now.